File: Desktop/TeamRes-seaport-final/dbms/backend/routes/immigration.py
import os, random, string
from flask import Blueprint, request, send_from_directory
from database import get_db, row_to_dict, rows_to_list, api_response, api_error
from werkzeug.utils import secure_filename

immigration_bp = Blueprint('immigration', __name__)

UPLOADS_DIR = os.path.join(os.path.dirname(__file__), '..', 'uploads', 'photos')
os.makedirs(UPLOADS_DIR, exist_ok=True)

# The demo passport data for Vikram Singh
DEMO_PASSPORT = {
    'passport_no':  'Z8892104',
    'name':         'Vikram Singh',
    'nationality':  'Indian',
    'dob':          '1988-03-15',
    'gender':       'Male',
    'place_of_birth': 'New Delhi',
    'date_of_issue':  '2021-01-12',
    'date_of_expiry': '2031-01-11',
    'mrz_line1': 'P<INDSINGH<<VIKRAM<<<<<<<<<<<<<<<<<<<<<<<<',
    'mrz_line2': 'Z88921048IND8803154M3101118<<<<<<<<<<<<<<<6'
}


@immigration_bp.route('/verify-passport', methods=['POST'])
def verify_passport():
    data     = request.get_json(silent=True) or {}
    pno      = data.get('passport_no', '').strip()
    ocr_name = data.get('ocr_name', '').strip()

    if not pno and not ocr_name:
        return api_error('passport_no or ocr_name required')

    db = get_db()
    try:
        row = None
        if pno:
            row = db.execute("SELECT * FROM entities WHERE passport_no=?", (pno,)).fetchone()
        if not row and ocr_name:
            row = db.execute(
                "SELECT * FROM entities WHERE name LIKE ?",
                (f'%{ocr_name.split()[0]}%',)
            ).fetchone()
    finally:
        db.close()

    if row:
        r = dict(row)
        face_match = 96 if r['passport_no'] == 'Z8892104' else (
            0 if r['is_blacklist'] else
            (int(85 + (100 - r['risk_score']) * 0.1))
        )
        checks = {
            'mrz_valid':        r['is_blacklist'] == 0,
            'not_expired':      True,
            'watchlist_clear':  r['is_blacklist'] == 0,
            'interpol_clear':   r['is_blacklist'] == 0,
            'face_match_score': face_match,
        }
        overall = 'Verified' if all(checks[k] for k in checks if k != 'face_match_score') and face_match > 80 else 'Flagged'
        return api_response(data={
            'found':          True,
            'entity':         r,
            'checks':         checks,
            'overall_status': overall,
            'is_blacklist':   bool(r['is_blacklist']),
            'blacklist_reason': r.get('blacklist_reason')
        })

    return api_response(data={
        'found':          False,
        'entity':         None,
        'checks': {
            'mrz_valid':        True,
            'not_expired':      True,
            'watchlist_clear':  True,
            'interpol_clear':   True,
            'face_match_score': 72,
        },
        'overall_status': 'Pending'
    }, message='Document scanned — manual review recommended')


@immigration_bp.route('/travelers', methods=['GET'])
def search_travelers():
    q      = request.args.get('q', '').strip()
    status = request.args.get('status', '')
    nat    = request.args.get('nationality', '')
    limit  = min(int(request.args.get('limit', 50)), 200)
    offset = int(request.args.get('offset', 0))

    conditions = ["type='Traveler'"]
    params     = []
    if q:
        conditions.append("(name LIKE ? OR passport_no LIKE ?)")
        params += [f'%{q}%', f'%{q}%']
    if status:
        conditions.append("status=?")
        params.append(status)
    if nat:
        conditions.append("nationality LIKE ?")
        params.append(f'%{nat}%')

    where = ' AND '.join(conditions)
    db = get_db()
    try:
        rows  = db.execute(
            f"SELECT * FROM entities WHERE {where} ORDER BY created_at DESC LIMIT ? OFFSET ?",
            params + [limit, offset]
        ).fetchall()
        total = db.execute(
            f"SELECT COUNT(*) as c FROM entities WHERE {where}", params
        ).fetchone()['c']
    finally:
        db.close()

    return api_response(data={'items': [dict(r) for r in rows], 'total': total})


@immigration_bp.route('/travelers/<entity_id>', methods=['GET'])
def get_traveler(entity_id):
    """Fetch a single traveler by ID (used by the Edit modal)."""
    db = get_db()
    try:
        row = db.execute(
            "SELECT * FROM entities WHERE id=? AND type='Traveler'", (entity_id,)
        ).fetchone()
    finally:
        db.close()
    if not row:
        return api_error('Traveler not found', 404)
    return api_response(data=dict(row))


@immigration_bp.route('/travelers', methods=['POST'])
def add_traveler():
    """Create a new Traveler entity."""
    data = request.get_json(silent=True) or {}
    required = ['name', 'nationality']
    for f in required:
        if not data.get(f):
            return api_error(f'Field required: {f}')

    entity_id   = f"BMS-TRV-{random.randint(10000,99999)}"
    passport_no = data.get('passport_no') or f"TEMP-TRV-{random.randint(1000,9999)}"
    # Status drives blacklist flag: 'Blacklisted' status → is_blacklist=1
    status      = data.get('status', 'Under Verification')
    blacklisted = 1 if (data.get('blacklisted') or status == 'Blacklisted') else 0

    db = get_db()
    try:
        db.execute("""
            INSERT INTO entities
              (id, name, passport_no, nationality, type, entry_point, status,
               risk_score, is_blacklist, dob, gender, visit_reason, visa_status,
               created_at, updated_at)
            VALUES (?,?,?,?,?,?,?,?,?,?,?,?,?,datetime('now'),datetime('now'))
        """, (
            entity_id, data['name'], passport_no, data['nationality'],
            'Traveler',
            data.get('entry_point', ''),
            status,
            int(data.get('risk_score', 0)),
            blacklisted,
            data.get('dob', ''),
            data.get('gender', ''),
            data.get('visit_reason', ''),
            data.get('visa_status', 'None'),
        ))
        db.commit()
    except Exception as e:
        return api_error(f'Database error: {str(e)}', 500)
    finally:
        db.close()

    return api_response(data={'id': entity_id, 'passport_no': passport_no},
                        message='Traveler added successfully')


@immigration_bp.route('/travelers/<entity_id>', methods=['PATCH'])
def update_traveler(entity_id):
    """Update any field on a Traveler entity."""
    data = request.get_json(silent=True) or {}
    allowed = ['name', 'passport_no', 'nationality', 'gender', 'dob',
               'entry_point', 'visit_reason', 'status', 'visa_status',
               'is_blacklist', 'blacklist_reason', 'risk_score']

    # Map 'blacklisted' convenience field
    if 'blacklisted' in data:
        data['is_blacklist'] = 1 if data.pop('blacklisted') else 0
    # Status drives blacklist flag: setting status='Blacklisted' also sets is_blacklist=1
    if 'status' in data:
        if data['status'] == 'Blacklisted':
            data['is_blacklist'] = 1
        elif data['status'] in ('Verified', 'Under Verification', 'Pending'):
            data.setdefault('is_blacklist', 0)

    fields = {k: v for k, v in data.items() if k in allowed}
    if not fields:
        return api_error('No valid fields provided to update')

    set_clause = ', '.join(f"{k}=?" for k in fields)
    values     = list(fields.values()) + [entity_id]

    db = get_db()
    try:
        db.execute(
            f"UPDATE entities SET {set_clause}, updated_at=datetime('now') WHERE id=? AND type='Traveler'",
            values
        )
        db.commit()
        changed = db.execute("SELECT changes() as n").fetchone()['n']
        if changed == 0:
            return api_error('Traveler not found', 404)
    except Exception as e:
        return api_error(str(e), 500)
    finally:
        db.close()

    return api_response(message='Traveler updated successfully')


@immigration_bp.route('/travelers/<entity_id>', methods=['DELETE'])
def delete_traveler(entity_id):
    """Delete a Traveler entity."""
    db = get_db()
    try:
        db.execute(
            "DELETE FROM entities WHERE id=? AND type='Traveler'", (entity_id,)
        )
        db.commit()
        changed = db.execute("SELECT changes() as n").fetchone()['n']
        if changed == 0:
            return api_error('Traveler not found', 404)
    except Exception as e:
        return api_error(str(e), 500)
    finally:
        db.close()

    return api_response(message='Traveler deleted')


@immigration_bp.route('/travelers/<entity_id>/photo', methods=['POST'])
def upload_photo(entity_id):
    """Store a passport photo associated with the traveler entity."""
    if 'photo' not in request.files:
        return api_error('No photo file provided')

    photo = request.files['photo']
    if not photo.filename:
        return api_error('Empty filename')

    ext      = os.path.splitext(secure_filename(photo.filename))[1].lower() or '.jpg'
    filename = f"{entity_id}{ext}"
    save_path = os.path.join(UPLOADS_DIR, filename)
    photo.save(save_path)

    # Store relative path reference in passport_photo column
    photo_url = f"/uploads/photos/{filename}"
    db = get_db()
    try:
        db.execute(
            "UPDATE entities SET passport_photo=?, updated_at=datetime('now') WHERE id=?",
            (photo_url, entity_id)
        )
        db.commit()
    finally:
        db.close()

    return api_response(data={'photo_url': photo_url}, message='Photo uploaded successfully')


@immigration_bp.route('/grant-entry', methods=['POST'])
def grant_entry():
    data = request.get_json(silent=True) or {}
    passport_no = data.get('passport_no', '')
    if not passport_no:
        return api_error('passport_no required')

    db = get_db()
    try:
        db.execute(
            "UPDATE entities SET status='Verified', updated_at=datetime('now') WHERE passport_no=?",
            (passport_no,)
        )
        db.commit()
    finally:
        db.close()

    return api_response(message=f'Entry granted for passport {passport_no}')

File: Desktop/TeamRes-seaport-final/dbms/frontend/js/immigration.js
/* immigration.js — OCR pipeline, scan animation, certificate render, Traveler CRUD */
/* NOTE FOR DEVELOPERS:
 * To open the traveler modal programmatically:
 *   openTravelerModal()              → Add mode (blank form)
 *   openTravelerModal('ENTITY_ID')   → Edit mode (pre-filled from API)
 *   deleteTraveler('ENTITY_ID', 'Name') → Delete with confirm
 *   loadTravelers(query, status)     → Refresh table
 */

'use strict';

// ── State ────────────────────────────────────────────────────────
let _webcamStream = null;
let _selectedFile = null;
let _modalMode    = 'add';   // 'add' | 'edit'
let _editId       = null;
let _photoFile    = null;    // selected photo in modal

// ── Webcam ───────────────────────────────────────────────────────
async function startCamera() {
  try {
    const stream = await navigator.mediaDevices.getUserMedia({
      video: { width: { ideal: 1280 }, height: { ideal: 720 }, facingMode: 'user' }
    });
    const video = document.getElementById('webcam-feed');
    if (video) { video.srcObject = stream; _webcamStream = stream; }
    const dot = document.getElementById('camera-dot');
    if (dot) dot.style.opacity = '1';
    const label = document.getElementById('camera-label');
    if (label) label.textContent = '● Camera Active';
  } catch {
    const label = document.getElementById('camera-label');
    if (label) label.textContent = '⚠ Camera unavailable';
  }
}

function captureFrame() {
  const video = document.getElementById('webcam-feed');
  if (!video || !video.videoWidth) { showToast('Camera not active', 'warning'); return; }
  const canvas = document.createElement('canvas');
  canvas.width  = video.videoWidth;
  canvas.height = video.videoHeight;
  canvas.getContext('2d').drawImage(video, 0, 0);
  showToast('Frame captured — face image stored for match analysis', 'info');
}

// ── Dropzone ─────────────────────────────────────────────────────
function setupDropzone() {
  const dz  = document.getElementById('dropzone');
  const inp = document.getElementById('file-input');
  if (!dz || !inp) return;

  dz.addEventListener('click', () => inp.click());
  inp.addEventListener('change', () => handleFile(inp.files[0]));
  dz.addEventListener('dragover', e => { e.preventDefault(); dz.classList.add('drag-over'); });
  dz.addEventListener('dragleave', () => dz.classList.remove('drag-over'));
  dz.addEventListener('drop', e => {
    e.preventDefault();
    dz.classList.remove('drag-over');
    if (e.dataTransfer.files.length) handleFile(e.dataTransfer.files[0]);
  });
}

function handleFile(file) {
  if (!file) return;
  _selectedFile = file;
  const dz = document.getElementById('dropzone');
  const scanBtn = document.getElementById('btn-scan');
  if (dz) {
    dz.classList.add('has-file');
    dz.innerHTML = `<div style="padding:8px;text-align:center">
      <div style="font-size:13px;font-weight:600;color:var(--color-success)">✓ ${file.name}</div>
      <div style="font-size:11px;color:var(--color-text-muted);margin-top:4px">${(file.size/1024).toFixed(1)} KB — Ready to scan</div>
    </div>`;
  }
  if (scanBtn) scanBtn.disabled = false;
}

// ── Scan sequence ─────────────────────────────────────────────────
const SCAN_STEPS = [
  [0,    'Initializing OCR engine...'],
  [700,  'Extracting biometric fields from document...'],
  [1600, 'Cross-referencing passport database...'],
  [2400, 'Running face match analysis...'],
  [3100, 'Generating verification report...'],
];

async function initiateScan() {
  if (!_selectedFile) { showToast('Please upload a passport document first', 'error'); return; }
  const btn = document.getElementById('btn-scan');
  if (btn) { btn.disabled = true; btn.textContent = 'Scanning...'; }

  const scanContainer = document.getElementById('scan-container');
  if (scanContainer) {
    const overlay = document.createElement('div');
    overlay.className = 'scan-overlay';
    overlay.innerHTML = '<div class="laser-line"></div>';
    scanContainer.style.position = 'relative';
    scanContainer.appendChild(overlay);
    setTimeout(() => overlay.remove(), 3000);
  }

  const statusEl = document.getElementById('scan-status');
  SCAN_STEPS.forEach(([delay, msg]) => setTimeout(() => { if (statusEl) statusEl.textContent = msg; }, delay));

  await new Promise(r => setTimeout(r, 3500));

  const filename = _selectedFile.name.toLowerCase();
  const passportNo = (filename.includes('vikram') || filename.includes('passport') || filename.includes('mockup'))
    ? 'Z8892104' : null;

  const res = await apiFetch('/api/immigration/verify-passport', {
    method: 'POST',
    body: JSON.stringify({ passport_no: passportNo || '', ocr_name: passportNo ? '' : 'Unknown' })
  });

  if (res.success) {
    renderCertificate(res.data);
    if (statusEl) statusEl.textContent = '✓ Verification complete.';
  } else {
    showToast('Verification API error: ' + (res.message || 'Unknown error'), 'error');
  }
  if (btn) { btn.disabled = false; btn.textContent = 'Initiate Verification Scan →'; }
}

function renderCertificate(data) {
  const container = document.getElementById('cert-container');
  if (!container) return;
  const e = data.entity || {};
  const checks = data.checks || {};
  const overall = data.overall_status || 'Pending';
  const isVerified = overall === 'Verified';
  const isBlacklist = data.is_blacklist;
  const score = checks.face_match_score || 0;

  const checkRows = [
    ['MRZ Validation',       checks.mrz_valid],
    ['Passport Not Expired', checks.not_expired],
    ['Watchlist Clear',      checks.watchlist_clear],
    ['INTERPOL Clear',       checks.interpol_clear],
  ].map(([label, ok]) => `
    <div class="cert-check-row">
      <span class="${ok ? 'check-ok' : 'check-fail'}">${ok ? '✓' : '✕'}</span>
      <span>${label}</span>
    </div>`).join('');

  container.innerHTML = `
    <div class="cert-panel ${isBlacklist ? 'flagged' : isVerified ? 'verified' : 'flagged'}">
      <div class="cert-header ${isBlacklist ? 'flagged' : isVerified ? 'verified' : 'flagged'}">
        ${isVerified && !isBlacklist ? '✓ VERIFICATION SUCCESSFUL' : '⚠ VERIFICATION ALERT'}
        ${isBlacklist ? ' — BLACKLISTED ENTITY — DO NOT GRANT ENTRY' : ''}
      </div>
      <div class="cert-body">
        <div style="display:grid;grid-template-columns:auto 1fr;gap:20px;align-items:start;margin-bottom:16px">
          <div style="text-align:center">
            <div class="cert-score" style="color:${isVerified&&!isBlacklist?'var(--color-success)':'var(--color-alert)'}">${score}%</div>
            <div style="font-size:11px;color:var(--color-text-muted);margin-top:4px">Face Match</div>
          </div>
          <div class="data-grid">
            <div class="data-field"><span class="data-label">Name</span><span class="data-value">${e.name||'—'}</span></div>
            <div class="data-field"><span class="data-label">Passport No.</span><span class="data-value mono">${e.passport_no||'—'}</span></div>
            <div class="data-field"><span class="data-label">Nationality</span><span class="data-value">${e.nationality||'—'}</span></div>
            <div class="data-field"><span class="data-label">Date of Birth</span><span class="data-value">${e.dob||'—'}</span></div>
          </div>
        </div>
        <div class="cert-checks">${checkRows}</div>
        ${isBlacklist ? `<div class="alert-banner error" style="margin-top:14px"><div><div class="alert-title">⚠ BLACKLIST ALERT</div><div class="alert-body-text">${data.blacklist_reason||'Entity is on the watchlist'}</div></div></div>` : ''}
        ${!isBlacklist && isVerified
          ? `<div style="margin-top:16px;display:flex;gap:8px;align-items:center">
               <button class="btn btn-success" onclick="grantEntry('${e.passport_no}')">✓ Grant Entry</button>
             </div>`
          : `<button class="btn btn-danger" style="margin-top:14px">Deny Entry — Escalate</button>`}
      </div>
    </div>`;
  container.scrollIntoView({ behavior: 'smooth', block: 'nearest' });
}

async function grantEntry(passportNo) {
  const res = await apiFetch('/api/immigration/grant-entry', {
    method: 'POST', body: JSON.stringify({ passport_no: passportNo })
  });
  if (res.success) showToast(`Entry granted for passport ${passportNo}`, 'success');
  else showToast('Grant entry failed: ' + res.message, 'error');
}

// ── Traveler DB Table ─────────────────────────────────────────────
async function loadTravelers(q = '', status = '') {
  const tbodyEl = document.getElementById('traveler-tbody');
  if (!tbodyEl) return;
  tbodyEl.innerHTML = '<tr><td colspan="10" style="text-align:center;padding:20px;color:var(--color-text-muted)">Loading...</td></tr>';

  const url = `/api/immigration/travelers?q=${encodeURIComponent(q)}&status=${encodeURIComponent(status)}&limit=50`;
  const res = await apiFetch(url);
  if (!res.success) {
    showToast('Failed to load travelers: ' + (res.message || 'error'), 'error');
    tbodyEl.innerHTML = '<tr><td colspan="10" style="text-align:center;padding:20px;color:var(--color-alert)">Error loading travelers</td></tr>';
    return;
  }

  const items = res.data.items;
  if (!items.length) {
    tbodyEl.innerHTML = '<tr><td colspan="10" style="text-align:center;padding:24px;color:var(--color-text-muted)">No travelers found</td></tr>';
    return;
  }

  tbodyEl.innerHTML = items.map(r => {
    const photoCell = r.passport_photo
      ? `<a href="${r.passport_photo}" target="_blank" title="View photo"><img src="${r.passport_photo}" class="photo-thumb" alt="Photo" style="cursor:pointer"></a>`
      : `<div class="photo-placeholder" title="No photo">👤</div>`;
    const isBlacklisted = r.status === 'Blacklisted' || r.is_blacklist;
    return `<tr style="${isBlacklisted ? 'background:#FEF2F2' : ''}">
      <td style="text-align:center">${photoCell}</td>
      <td class="font-mono">${r.passport_no || '—'}</td>
      <td><strong>${r.name}</strong></td>
      <td>${r.nationality}</td>
      <td>${r.gender || '—'}</td>
      <td>${r.dob || '—'}</td>
      <td>${(r.entry_point || '').split(',')[0] || '—'}</td>
      <td>${r.visit_reason || '—'}</td>
      <td>${r.visa_status || '—'}</td>
      <td>${statusBadge(isBlacklisted ? 'Blacklisted' : (r.status || 'Under Verification'))}</td>
      <td>
        <div style="display:flex;gap:4px">
          <button class="btn btn-secondary btn-sm" onclick="openTravelerModal('${r.id}')">✏ Edit</button>
          <button class="btn btn-danger btn-sm" onclick="deleteTraveler('${r.id}', ${JSON.stringify(r.name)})">🗑</button>
        </div>
      </td>
    </tr>`;
  }).join('');
}

// ── Modal: Open ───────────────────────────────────────────────────
async function openTravelerModal(entityId) {
  _modalMode = entityId ? 'edit' : 'add';
  _editId    = entityId || null;
  _photoFile = null;

  // Reset all form fields
  _setField('m-name', '');
  _setField('m-nationality', '');
  _setField('m-passport', '');
  _setField('m-dob', '');
  _setField('m-entry-point', '');
  _setField('m-visit-reason', '');
  _setSelect('m-gender', '');
  _setSelect('m-visa-status', 'None');
  _setSelect('m-status', 'Under Verification');
  const blEl = document.getElementById('m-blacklisted');
  if (blEl) blEl.checked = false;

  // Photo section always visible (both add and edit)
  const photoGroup = document.getElementById('photo-upload-group');
  if (photoGroup) photoGroup.style.display = '';
  _clearPhotoPreview();

  // Update modal title
  const titleEl = document.getElementById('modal-title');
  if (titleEl) titleEl.textContent = entityId ? 'Edit Traveler' : 'Add Traveler';
  const saveBtn = document.getElementById('modal-save-btn');
  if (saveBtn) saveBtn.textContent = entityId ? 'Save Changes' : 'Add Traveler';

  // Show modal immediately (load data in background for edit mode)
  const modal = document.getElementById('traveler-modal');
  if (modal) modal.style.display = 'flex';

  // If edit mode, fetch and fill the data
  if (entityId) {
    if (saveBtn) { saveBtn.disabled = true; saveBtn.textContent = 'Loading...'; }
    const res = await apiFetch(`/api/immigration/travelers/${entityId}`);
    if (res.success) {
      const traveler = res.data;

      if (traveler) {
        _setField('m-name', traveler.name || '');
        _setField('m-nationality', traveler.nationality || '');
        _setField('m-passport', traveler.passport_no || '');
        _setField('m-dob', traveler.dob || '');
        _setField('m-entry-point', traveler.entry_point || '');
        _setField('m-visit-reason', traveler.visit_reason || '');
        _setSelect('m-gender', traveler.gender || '');
        _setSelect('m-visa-status', traveler.visa_status || 'None');
        _setSelect('m-status', traveler.status || 'Under Verification');

        if (traveler.passport_photo) {
          const thumb = document.getElementById('modal-photo-thumb');
          if (thumb) { thumb.src = traveler.passport_photo; thumb.style.display = 'block'; }
          const pname = document.getElementById('modal-photo-name');
          if (pname) pname.textContent = 'Current passport photo';
        }
      } else {
        showToast('Traveler record not found', 'error');
        closeTravelerModal();
        return;
      }
    }
    if (saveBtn) { saveBtn.disabled = false; saveBtn.textContent = 'Save Changes'; }
  }
}

function _setField(id, value) {
  const el = document.getElementById(id);
  if (el) el.value = value;
}

function _setSelect(id, value) {
  const el = document.getElementById(id);
  if (!el) return;
  // Try to set the value directly
  el.value = value;
  // If no option matched, default to first
  if (el.value !== value) el.selectedIndex = 0;
}

function _clearPhotoPreview() {
  const thumb = document.getElementById('modal-photo-thumb');
  if (thumb) { thumb.src = ''; thumb.style.display = 'none'; }
  const pname = document.getElementById('modal-photo-name');
  if (pname) pname.textContent = '';
  const fileInput = document.getElementById('m-photo');
  if (fileInput) fileInput.value = '';
  _photoFile = null;
}

// ── Modal: Close ──────────────────────────────────────────────────
function closeTravelerModal(e) {
  // If called from the overlay click, only close if the overlay itself was clicked
  if (e && e.type === 'click') {
    const overlay = document.getElementById('traveler-modal');
    if (e.target !== overlay) return;
  }
  const modal = document.getElementById('traveler-modal');
  if (modal) modal.style.display = 'none';
  _photoFile = null;
}

// ── Modal: Save (Add or Edit) ─────────────────────────────────────
async function saveTraveler() {
  const name        = (document.getElementById('m-name')?.value || '').trim();
  const nationality = (document.getElementById('m-nationality')?.value || '').trim();
  if (!name)        { showToast('Full Name is required', 'error'); document.getElementById('m-name')?.focus(); return; }
  if (!nationality) { showToast('Nationality is required', 'error'); document.getElementById('m-nationality')?.focus(); return; }

  const payload = {
    name,
    nationality,
    passport_no:  (document.getElementById('m-passport')?.value || '').trim(),
    gender:       document.getElementById('m-gender')?.value || '',
    dob:          document.getElementById('m-dob')?.value || '',
    entry_point:  (document.getElementById('m-entry-point')?.value || '').trim(),
    visit_reason: (document.getElementById('m-visit-reason')?.value || '').trim(),
    visa_status:  document.getElementById('m-visa-status')?.value || 'None',
    // Status drives is_blacklist automatically
    status:       document.getElementById('m-status')?.value || 'Under Verification',
    blacklisted:  document.getElementById('m-status')?.value === 'Blacklisted',
  };

  const saveBtn = document.getElementById('modal-save-btn');
  if (saveBtn) { saveBtn.disabled = true; saveBtn.textContent = 'Saving...'; }

  let res;
  let entityId = _editId;

  if (_modalMode === 'edit' && _editId) {
    // PATCH to update existing
    res = await apiFetch(`/api/immigration/travelers/${_editId}`, {
      method: 'PATCH',
      body: JSON.stringify(payload)
    });
  } else {
    // POST to create new
    res = await apiFetch('/api/immigration/travelers', {
      method: 'POST',
      body: JSON.stringify(payload)
    });
    if (res.success) entityId = res.data?.id;
  }

  if (saveBtn) { saveBtn.disabled = false; saveBtn.textContent = _modalMode === 'edit' ? 'Save Changes' : 'Add Traveler'; }

  if (!res.success) {
    showToast('Save failed: ' + (res.message || 'Unknown error'), 'error');
    return;
  }

  // Upload photo if one was selected
  if (_photoFile && entityId) {
    const uploaded = await uploadPassportPhoto(entityId, _photoFile);
    if (!uploaded) showToast('Traveler saved but photo upload failed', 'warning');
  }

  showToast(_modalMode === 'edit' ? 'Traveler updated successfully' : 'Traveler added successfully', 'success');
  const modal = document.getElementById('traveler-modal');
  if (modal) modal.style.display = 'none';

  // Refresh table
  loadTravelers(
    document.getElementById('traveler-search')?.value || '',
    document.getElementById('traveler-status')?.value || ''
  );
}

// ── Delete ────────────────────────────────────────────────────────
async function deleteTraveler(entityId, name) {
  if (!confirm(`Delete traveler "${name}"?\n\nThis cannot be undone.`)) return;

  const res = await apiFetch(`/api/immigration/travelers/${entityId}`, { method: 'DELETE' });
  if (res.success) {
    showToast('Traveler deleted', 'success');
    loadTravelers(
      document.getElementById('traveler-search')?.value || '',
      document.getElementById('traveler-status')?.value || ''
    );
  } else {
    showToast('Delete failed: ' + (res.message || 'error'), 'error');
  }
}

// ── Photo upload (multipart — does NOT use apiFetch to avoid wrong Content-Type) ──
async function uploadPassportPhoto(entityId, file) {
  const fd = new FormData();
  fd.append('photo', file);
  try {
    // Do NOT set Content-Type header — browser sets it automatically with boundary for multipart
    const raw = await fetch(`/api/immigration/travelers/${entityId}/photo`, {
      method: 'POST',
      body: fd
    });
    const json = await raw.json();
    if (!json.success) { showToast('Photo upload failed: ' + json.message, 'error'); return false; }
    showToast('Passport photo saved', 'success');
    return true;
  } catch (e) {
    showToast('Photo upload error: ' + e.message, 'error');
    return false;
  }
}

// ── DOMContentLoaded ──────────────────────────────────────────────
document.addEventListener('DOMContentLoaded', () => {
  startCamera();
  setupDropzone();

  document.getElementById('btn-scan')?.addEventListener('click', initiateScan);
  document.getElementById('btn-capture')?.addEventListener('click', captureFrame);

  // Photo file input in modal — preview on selection
  document.getElementById('m-photo')?.addEventListener('change', function () {
    const file = this.files?.[0];
    if (!file) return;
    _photoFile = file;
    document.getElementById('modal-photo-name').textContent = file.name;
    const thumb = document.getElementById('modal-photo-thumb');
    if (thumb) {
      const reader = new FileReader();
      reader.onload = ev => { thumb.src = ev.target.result; thumb.style.display = 'block'; };
      reader.readAsDataURL(file);
    }
  });

  // Tab switching
  document.querySelectorAll('.tab-item').forEach(tab => {
    tab.addEventListener('click', () => {
      document.querySelectorAll('.tab-item').forEach(t => t.classList.remove('active'));
      document.querySelectorAll('.tab-content').forEach(c => c.classList.remove('active'));
      tab.classList.add('active');
      const content = document.getElementById(tab.dataset.tab);
      if (content) content.classList.add('active');
      if (tab.dataset.tab === 'tab-travelers') {
        loadTravelers();
      }
    });
  });

  // Traveler search/filter
  document.getElementById('traveler-search')?.addEventListener('input', function () {
    loadTravelers(this.value, document.getElementById('traveler-status')?.value || '');
  });
  document.getElementById('traveler-status')?.addEventListener('change', function () {
    loadTravelers(document.getElementById('traveler-search')?.value || '', this.value);
  });

  // Modal overlay click-outside-to-close
  document.getElementById('traveler-modal')?.addEventListener('click', closeTravelerModal);
});
